cxx-tests/src/cxx_support_test_cpp.h

```diff
@@ -18,35 +18,155 @@
 #define CXX_SUPPORT_TEST_CPP
 
 #include <cstdint>
-#include <cstring>
 #include <memory>
+#include <cstddef>
+#include <new>
+#include <type_traits>
+#include <utility>
 
-constexpr uint8_t kUninitialized = 0;
-constexpr uint8_t kInitialized = 1;
-constexpr uint8_t kMethodCalled = 2;
+#include "moveit-cxx-tests/src/tests.rs.h"
+
+// Test code status reporting
+
+extern Status status;
+
+inline Status get_status() {
+  return status;
+}
+
+inline void set_status(Status new_status) {
+  status = new_status;
+}
+
+inline void reset_status() {
+  set_status(Status::Unallocated);
+}
+
+// Actual classes that would really be in production
+// code
 
 class Foo {
  public:
-  Foo() { data[0] = kInitialized; }
-  uint8_t get_status() const { return data[0]; }
-  void modify() { data[0] = kMethodCalled; }
+  Foo() {
+    data[0] = 0;
+    set_status(Status::Initialized);
+  }
+  bool get_a() const { return data[0] != 0; }
+  void modify() { data[0] = 1; }
+  ~Foo() {
+    set_status(Status::Destructed);
+  }
+  // We test using a custom operator new/delete
+  // pair partly for purposes of recording status,
+  // and partly so that we can verify that moveit
+  // will work for C++ classes which do similar.
+  void* operator new(size_t count) {
+    set_status(Status::Allocated);
+    return ::operator new(count);
+  }
+  void* operator new(size_t count, void* ptr) {
+    set_status(Status::Allocated);
+    return ::operator new(count, ptr);
+  }
+  void operator delete(void* ptr) noexcept {
+    ::operator delete(ptr);
+    set_status(Status::Deallocated);
+  }
 
  private:
-  uint32_t data[4];  // exactly match layout declared in Rust.
+  uint32_t data[4];
 };
 
-inline Foo* CreateUninitializedFoo() {
-  std::allocator<Foo> alloc;
-  Foo* data = alloc.allocate(1);
-  std::memcpy(data, &kUninitialized, 1);
-  return data;
+class Bar {
+public:
+  Bar() {
+    data[0] = 0;
+    set_status(Status::Initialized);
+  }
+  ~Bar() {
+    set_status(Status::Destructed);
+  }
+  void do_nothing() const {}
+
+private:
+  uint32_t data[4];
+};
+
+// Everything below here would be generated by some code generator
+
+// Mechanics to call custom deallocators
+
+template<class T>
+auto delete_imp(T* ptr, int) -> decltype(T::operator delete(ptr), void())
+{
+  T::operator delete(ptr);
+}
+
+template<class T>
+auto delete_imp(T* ptr, long) -> decltype(::operator delete(ptr), void())
+{
+  ::operator delete(ptr);
 }
 
-inline void FreeUninitializedFoo(Foo* foo) {
-  std::allocator<Foo> alloc;
-  alloc.deallocate(foo, 1);
+template<class T>
+auto delete_appropriately(T* obj) -> decltype(delete_imp(obj, 0), void())
+{
+  // 0 is a better match for the first 'delete_imp' so will match
+  // preferentially.
+  delete_imp(obj, 0);
+}
+
+template<class T>
+auto new_imp(size_t count, int, T*) -> decltype(T::operator new(count))
+{
+  return T::operator new(count);
+}
+
+template<class T>
+auto new_imp(size_t count, long, T*) -> decltype(::operator new(count))
```

Comment:
This `decltype` will never trigger an error, so I'd just declare this function as returning `size_t`.